### PR TITLE
Implement Natural/subtract

### PR DESCRIPTION
**Pull request by ocharles**
... as standardized in https://github.com/dhall-lang/dhall-lang/pull/650

### CODE REVIEW

**Comment by AndrasKovacs** [comment]
The [commented-out](https://github.com/dhall-lang/dhall-haskell/blob/8ba65f55f22dbe4c19a075a34a2ebbcb175f21ac/dhall/src/Dhall/Eval.hs#L811) `qApp` should be there.

**Comment by ocharles** [comment]
Oh, it wasn't curried out before but it still fails to terminate with it. Let me push that.

**Comment by ocharles** [comment]
Ok, pushed that corrected version (but it still gets stuck)

**Comment by f-f** [comment]
@ocharles shouldn't `NaturalTruncatedSubtract` constructor be more like `NaturalPlus` (i.e. containing two values) than like `NaturalShow`? (with no values attached)

**Comment by AndrasKovacs** [comment]
@f-f yeah that would work. It's also possible to do the curried version (I'm going to paste here that in a minute), but if partial application is not possible in the syntax, there is no reason to define `NaturalTruncatedSubtract` as curried.

**Comment by ocharles** [comment]
I believe it's because partial application should be allowed. All the other terms are infix operators, I believe.

**Comment by AndrasKovacs** [comment]
OK, give me some time plz I'm a bit confused why this loops

**Comment by ocharles** [comment]
Of course, no rush from my part!

**Comment by sjakobi** [comment]
I guess I'm a bit late to the discussion. But wouldn't a function `Natural -> Natural -> Optional Natural` be slightly more useful? I mean

```
2 - 1 = Some 1
1 - 1 = Some 0
1 - 2 = None Natural
```

You could then use `Optional/fold` to derive truncated subtraction.

Less-than and equals operations would be very straightforward to implement.

(I suspect this was considered before but decided against for some reason?)

**Comment by ocharles** [comment]
@sjakobi See https://github.com/dhall-lang/dhall-lang/issues/602#issuecomment-506717501 - I believe I'm implementing faithfully to the linked comment.

**Comment by AndrasKovacs** [comment]
@ocharles You forgot to add `NaturalTruncatedSubtract` to `prettyPrimitiveExpression` in `Pretty/Internal.hs`.

**Comment by ocharles** [comment]
Oh man, that's horrific. I knew it must have been a type class spinning somewhere, thanks for taking the time to dig into that for me!

**Comment by Gabriel439** [comment]
Yeah, the pretty-printing infinite loop is a common land-mine that people stumble upon when making contributions (myself included).  If anybody knows how to fix it I'll make the change

**Comment by joneshf** [comment]
One option is to remove the default and handle the 20 other cases explicitly.

**Comment by Gabriel439** [comment]
This pull request seems to have the `Lift`-related work.  Is that intentional?

**Comment by ocharles** [comment]
Sort of. I was lazy and didn't start from master. I could debase, but I assumed the Lift stuff would go through

**Comment by ocharles** [comment]
Note that I planned to merge them sequentially, so they will still be split into separate commits

**Comment by Gabriel439** [comment]
@ocharles: I still think we should unbundle the `Lift`-related changes from this pull request, otherwise we'll have to block this pull request on the other one being merged first

**Comment by ocharles** [comment]
Yep, will do. I was lazy and just thought that would have been merged by now, but it's turning out more complicated!

**Comment by ocharles** [comment]
Ok, rebased.

**Comment by sjakobi** [comment]
Can you update the `Arbitrary` instance in `Dhall.Test.QuickCheck`? Feel free to initially generate `NaturalSubtract`s with a higher frequency, but please set it to `1` before merging.

**Comment by ocharles** [comment]
@Gabriel439 good to merge now that the new version is out?

**Comment by ocharles** [comment]
Just remembered I need to bump the submodule, I'll do that before merging and make sure the tests pass.

**Comment by Gabriel439** [comment]
I think if you merge in `master` it will fix the test failure that you are running into

**Comment by ocharles** [comment]
@Gabriel439 Nope. I'm not sure `dhall-lang` `master` works with `dhall-haskell` yet.

**Comment by sjakobi** [comment]
@ocharles I think you should checkout https://github.com/dhall-lang/dhall-lang/commit/c7082d910d956bcedfdc51daae989659a2db67bd for `dhall-lang`. The change in the failing parser test is more recent.

I believe `isNormalized` needs a bit more work too. Maybe up the frequency of `NaturalSubtract`s a bit and increase the `maxSuccesses` for the quickcheck test.

**Comment by sjakobi** [comment]
> I believe `isNormalized` needs a bit more work too. Maybe up the frequency of `NaturalSubtract`s a bit and increase the `maxSuccesses` for the quickcheck test.

…at least temporarily so you can see the failing cases more reliably.

**Comment by ocharles** [comment]
Thanks for the tip @sjakobi! That did indeed reveal a problem. I think we're good to go now.

File: dhall/src/Dhall/Core.hs [hunk 1]
```diff
@@ -1823,6 +1837,7 @@ isNormalized e0 = loop (denote e0)
           App NaturalEven (NaturalLit _) -> False
           App NaturalOdd (NaturalLit _) -> False
           App NaturalShow (NaturalLit _) -> False
+          App (App NaturalSubtract (NaturalLit _)) (NaturalLit _) -> False
```
Comment (sjakobi) [inline dhall/src/Dhall/Core.hs]: I believe you'll also need to handle `subtract x 0` and `subtract 0 x` here.

Comment (ocharles) [reply dhall/src/Dhall/Core.hs]: Hmm, I can't get a test failure. Let me try in the repl.

Comment (ocharles) [reply dhall/src/Dhall/Core.hs]: Yea, you're right - added those rules.

Comment (sjakobi) [reply dhall/src/Dhall/Core.hs]: Cheers!

**Pull request merged**